src/crystal/scheduler.cr

```diff
@@ -192,38 +188,34 @@ class Crystal::Scheduler
     end
 
     def run_loop
+      @worker_fiber = Fiber.current
+
       spawn_stack_pool_collector
 
-      fiber_channel = self.fiber_channel
       loop do
         @lock.lock
 
         if runnable = @runnables.shift?
-          @runnables << Fiber.current
+          @runnables << worker_fiber
           @lock.unlock
           resume(runnable)
         else
           @sleeping = true
           @lock.unlock
-
           Crystal.trace :sched, "mt:sleeping"
-          fiber = Crystal.trace(:sched, "mt:slept") { fiber_channel.receive }
-
-          @lock.lock
-          @sleeping = false
-          @runnables << Fiber.current
-          @lock.unlock
-          resume(fiber)
+          Crystal.trace(:sched, "mt:slept") { ::Fiber.suspend }
         end
       end
     end
 
     def send_fiber(fiber : Fiber)
       @lock.lock
+      @runnables << fiber
+
       if @sleeping
-        fiber_channel.send(fiber)
```

Comment:
What is the effect on performance? Not involving a pipe could be pretty impactful. 

Another nice effect is that it removes the scenario when the pipe end up full for some reason.

Reply:
It entirely depends on the implementation of `EventLoop#interrupt`:

- **libevent**: calls `event_base_loopexit` that will enqueue an event. I'm not sure what it does internally, probably one of `eventfd`, `EVFILT_USER` or a pipe.

- **epoll**: writes to an `eventfd`;

- **kqueue**: triggers an user event when `EVFILT_USER` is supported, otherwise writes a to a pipe;

- **IOCP**: currently sends an UserAPC to the waiting thread, but will post a custom completion key after #15238.